src/writer.jl

```diff
@@ -1,398 +1,277 @@
-    reward_name::AbstractString
-    obs_name::AbstractString
+    s_var_name::String = "state"
+    a_var_name::String = "action"
+    o_var_name::String = "observation"
+    r_var_name::String = "reward"
 
-    initial_belief::Vector{Float64}
+    s_name::Function = normalize
+    a_name::Function = normalize
+    o_name::Function = normalize
 
-    #initial_belief::Vector{Float64} # belief over partially observed vars
+    pretty::Bool = false
+end
 
-    function POMDPXFile(file_name::AbstractString; description::AbstractString="",
-                    initial_belief::Vector{Float64}=Float64[])
+POMDPXFile(filename::String, pretty::Bool) = POMDPXFile(; filename=filename, pretty=pretty)
+POMDPXFile(filename::String) = POMDPXFile(filename, false)
+
+a_var_name( px::POMDPXFile) = px.a_var_name
+s_var_name( px::POMDPXFile) = "$(px.s_var_name)0"
+sp_var_name(px::POMDPXFile) = "$(px.s_var_name)1"
+o_var_name( px::POMDPXFile) = px.o_var_name
+r_var_name( px::POMDPXFile) = px.r_var_name
+
+function build_xml(p::POMDP, px::POMDPXFile)
+    n_states = length(states(p))
+    n_actions = length(actions(p))
+    n_obs = length(observations(p))
+    n_nodes = 14 + n_states + n_states * n_actions * n_obs + n_states * n_actions * 2
+    pbar = Progress(n_nodes; dt=0.01)
+
+    doc  = XMLDocument()
+    root = ElementNode("pomdpx")
+    root["version"] = 0.1
+    root["id"] = replace(px.filename, ".pomdpx" => "")
+    root["xmlns:xsi"] = "http://www.w3.org/2001/XMLSchema-instance"
+    root["xsi:noNamespaceSchemaLocation"] = "https://raw.githubusercontent.com/JuliaPOMDP/sarsop/master/doc/POMDPX/pomdpx.xsd"
+    setroot!(doc, root)
+
+    addelement!(root, "Description", px.description)
+    addelement!(root, "Discount", "$(discount(p))")
+    next!(pbar)
+
+    build_variables!(root, p, px, pbar)
+    build_initial_beliefs(root, p, px, pbar)
+    build_transitions!(root, p, px, pbar)
+    build_observations!(root, p, px, pbar)
+    build_rewards!(root, p, px, pbar)
+
+    return doc
+end
 
-        if isempty(description)
-            description = "This is a pomdpx file for a partially observable MDP"
-        end
+function build_variables!(root::Node, p::POMDP, px::POMDPXFile, pbar)
+    variables = ElementNode("Variable")
+    link!(root, variables)
+
+    states_node = ElementNode("StateVar")
+    states_node["vnamePrev"] = s_var_name(px)
+    states_node["vnameCurr"] = sp_var_name(px)
+    states_node["fullyObs" ] = "false"
+    link!(variables, states_node)
+
+    actions_node = ElementNode("ActionVar")
+    actions_node["vname"] = a_var_name(px)
+    link!(variables, actions_node)
+
+    obs_node = ElementNode("ObsVar")
+    obs_node["vname"] = o_var_name(px)
+    link!(variables, obs_node)
+
+    reward_node = ElementNode("RewardVar")
+    reward_node["vname"] = r_var_name(px)
+    link!(variables, reward_node)
+    next!(pbar)
+
+    if px.pretty
+        all_states = join(["s_$(px.s_name(s))" for s=ordered_states(p)], " ")
```

Comment:
Is there some reason all states *need* to start with s?

Reply:
So, my main motive for this was that I’m not certain that a state `tiger-left` and observation `tiger-left` won’t be confused by the POMDPX parser (from upstream `sarsop`). If you know whether this would be the case, then I’m happy to remove it, period.

I suppose I can also test that, too. 😂 I was just focused on other things when generating the XML. Additionally, I’ve noticed that there’s some bug that leads to repeating XML nodes, so I need to clean that up before testing whether the POMDPX parser can handle same-name states/actions/observations.

Reply:
I think it is a good idea to have the "s_" *by default*. I just think it should be implemented like this in the POMDPXFile type:
```julia
s_name::Function = s->"s_"*normalize(s)
```
that way, people can override it *if they want*.